lib/Sema/TypeCheckMacros.cpp

```diff
@@ -543,6 +543,112 @@ static Identifier makeIdentifier(ASTContext &ctx, std::nullptr_t) {
   return Identifier();
 }
 
+bool swift::diagnoseInvalidAttachedMacro(MacroRole role,
+                                         Decl *attachedTo) {
+  switch (role) {
+  case MacroRole::Expression:
+  case MacroRole::Declaration:
+  case MacroRole::CodeItem:
+    llvm_unreachable("Invalid macro role for attached macro");
+
+  case MacroRole::Accessor:
+    // Only var decls and subscripts have accessors.
+    if (isa<AbstractStorageDecl>(attachedTo) && !isa<ParamDecl>(attachedTo))
+      return false;
+
+    break;
+
+  case MacroRole::MemberAttribute:
+  case MacroRole::Member:
+    // Nominal types and extensions can have members.
+    if (isa<NominalTypeDecl>(attachedTo) || isa<ExtensionDecl>(attachedTo))
+      return false;
+
+    break;
+
+  case MacroRole::Peer:
+    // Peer macros are allowed on everything except parameters.
+    if (!isa<ParamDecl>(attachedTo))
+      return false;
+
+    break;
+
+  case MacroRole::Conformance:
+  case MacroRole::Extension:
+    // Only primary declarations of nominal types
+    if (isa<NominalTypeDecl>(attachedTo))
```

Comment:
I thought one would be allowed to put an extension macro on an extension?

Reply:
The only reason I banned extensions here is because I didn't implement it yet. This came up [in the review thread](https://forums.swift.org/t/se-0402-generalize-conformance-macros-as-extension-macros/65965/16) too. It should be very straightforward to generalize the places that assume an extension macro is applied to a `NominalTypeDecl`.